src/workers/jobs.js

```diff
@@ -183,55 +224,137 @@ export async function loadContactsFromDataWarehouse(job) {
+    if (jobEvent.organizationId) {
+      // now that we've saved them all, we delete everyone that is opted out locally
+      // doing this in one go so that we can get the DB to do the indexed cell matching
+      const optOutCellCount = await r.knex('campaign_contact')
+        .whereIn('cell', function optouts() {
+          this.select('cell').from('opt_out').where('organization_id', jobEvent.organizationId)
+        })
+        .where('campaign_id', jobEvent.campaignId)
+        .delete()
+      console.log('OPTOUT CELL COUNT', optOutCellCount)
+    }
+    await r.table('job_request').get(jobEvent.jobId).delete()
+    return { 'completed': 1 }
+  } else if (jobEvent.part < (jobEvent.totalParts - 1)) {
+    const newPart = jobEvent.part + 1
+    const newJob = {
+      ...jobEvent,
+      part: newPart,
+      offset: newPart * jobEvent.step,
+      limit: jobEvent.step,
+      command: 'loadContactsFromDataWarehouseFragmentJob'
+    }
+    if (process.env.WAREHOUSE_DB_LAMBDA_ITERATION) {
+      console.log('SENDING TO LAMBDA loadContactsFromDataWarehouseFragment', newJob)
+      await sendJobToAWSLambda(newJob)
+      return { 'invokedAgain': 1 }
+    } else {
+      return loadContactsFromDataWarehouseFragment(newJob)
+    }
+  }
+}
+
+export async function loadContactsFromDataWarehouse(job) {
+  console.log('STARTING loadContactsFromDataWarehouse')
+  const jobMessages = []
+  const sqlQuery = job.payload
+  if (!sqlQuery.startsWith('SELECT') || sqlQuery.indexOf(';') >= 0) {
+    log.error('Malformed SQL statement.  Must begin with SELECT and not have any semicolons: ', sqlQuery)
+    return
+  }
+  if (!datawarehouse) {
+    log.error('No data warehouse connection, so cannot load contacts', job)
+    return
+  }
+
+  let knexCountRes
+  let knexCount
+  try {
+    knexCountRes = await datawarehouse.raw(`SELECT COUNT(*) FROM ( ${sqlQuery} )`)
+  } catch (err) {
+    log.error('Data warehouse count query failed: ', err)
+    jobMessages.push(`Error: Data warehouse count query failed: ${err}`)
+  }
+
+  if (knexCountRes) {
+    knexCount = knexCountRes.rows[0].count
+    console.log('WAREHOUSE COUNT', knexCount)
+  }
+  if (!knexCount) {
+    jobMessages.push('Error: Data warehouse query returned zero results')
+  }
+  const STEP = ((r.kninky && r.kninky.defaultsUnsupported)
+                ? 10 // sqlite has a max of 100 variables and ~8 or so are used per insert
+                : 10000) // default
   const campaign = await Campaign.get(job.campaign_id)
-  const optOutCellCount = await r.knex('campaign_contact')
-    .whereIn('cell', function () {
-      this.select('cell').from('opt_out').where('organization_id', campaign.organization_id)
-    })
-    .where('campaign_id', job.campaign_id)
-    .delete()
+  const totalParts = Math.ceil(knexCount / STEP)
 
-  console.log('OPTOUT CELL COUNT', optOutCellCount)
+  if (totalParts > 1 && /LIMIT/.test(sqlQuery)) {
```

Comment:
Let's add a comment here explaining why: "We're instrumenting this query with limit / offset to implement batching results back from this query, to get back currently scaling limitations. It'd be way more complicated to implement batching with a query that already does some batching itself, so we do the easier thing here and check whether queries contain LIMIT, and if so, give the user an error so they can rewrite this query to not preemptively batch."